x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/InitialSecurityConfigurationListener.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security;
+
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.DocWriteRequest;
+import org.elasticsearch.action.support.GroupedActionListener;
+import org.elasticsearch.action.support.WriteRequest;
+import org.elasticsearch.bootstrap.BootstrapInfo;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.common.settings.SecureString;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.index.engine.VersionConflictEngineException;
+import org.elasticsearch.xpack.core.XPackSettings;
+import org.elasticsearch.xpack.core.security.user.ElasticUser;
+import org.elasticsearch.xpack.core.ssl.SSLService;
+import org.elasticsearch.xpack.security.authc.esnative.NativeUsersStore;
+import org.elasticsearch.xpack.security.enrollment.InternalEnrollmentTokenGenerator;
+import org.elasticsearch.xpack.security.support.SecurityIndexManager;
+
+import java.io.PrintStream;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.BiConsumer;
+
+import static org.elasticsearch.xpack.security.authc.esnative.ReservedRealm.BOOTSTRAP_ELASTIC_PASSWORD;
+import static org.elasticsearch.xpack.security.tool.CommandUtils.generatePassword;
+
+public class InitialSecurityConfigurationListener implements BiConsumer<SecurityIndexManager.State, SecurityIndexManager.State> {
```

Comment:
This is tested in [packaging tests](https://github.com/elastic/elasticsearch/pull/78293/files#diff-e86b195280f586db0225e8873e6c60257a39aa64ed211da145a65e5e667a5c53R30) for now. Happy to discuss if we need unit tests and what they'd look like